src/lib/middleware/on-error.ts

```diff
@@ -0,0 +1,55 @@
+import { Middleware, AnyContext } from '../context'
+import { CallType } from '../call-types'
+
+type ErrorHandler = (error: Error, context: AnyContext) => any
+
+/**
+ * Patches emit function to intercept errors with a handler to be able to consume or map dispatched errors on stream, before existing listeners are invoked.
+ * Error events are passed on the handler and metadata are passed on re-thrown errors.
+ * @param emitter gRPC call
+ * @param handler Error handler that can either throw an error, that will be passed on or not throw, in which case cascade is stopped and call does not end with error
+ */
+const mapError = (emitter: any, handler: ErrorHandler) => {
+  const originalEmit = emitter.emit
+  const addMeta = (e: any) => {
+    // Trailing metadata on error calls are taken from `error.metadata`, we pass on the existing trailing metadata
+    // TODO: At this point existing metadata on error is lost
+    e.metadata = emitter.trailingMetadata
```

Comment:
Is it possible that - maybe - the emitter does not have the `trailingMetadata` yet? If we talking streams for example.

Reply:
At this point no, it is assigned along with the context before any of the handlers is invoked: https://github.com/grissius/protocat/blob/defa2a73b916543d7795d5039e93bba2a814b215/src/lib/server.ts#L79